### ISSUE TITLE
FastRoute falls back to GET on HEAD request

### ISSUE THREAD

**Issue by garnold** [slimphp/Slim#2386]
We were trying to figure out why `HEAD` requests from bots were making their way into our controllers. Then we discovered that [FastRoute falls back to `GET` on `HEAD` requests](https://github.com/nikic/FastRoute/blob/master/src/Dispatcher/RegexBasedAbstract.php#L33).

This is unexpected.

Related, I've noticed a number of issues where Slim users are requesting an `$app->head(...)` shortcut be added to Slim, and the response is often to use `$app->map(['HEAD'], ...)` instead. But given FastRoute's fallback you can just use `$app->get(...)` to handle `HEAD` requests. So my guess is this fallback behavior isn't well known in the Slim community.

**Comment by geggleto** [comment]
Hmm, HEAD should be GET just w/o a response body.

**Comment by damianopetrungaro** [comment]
As @geggleto said: https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.4

**Comment by akrabat** [comment]
Do we correctly ditch the body when it's a HEAD?

**Comment by designermonkey** [comment]
I'm not sure we do from a quick glance. We have the `isEmpty` method on a Response, but I can't see where we would check the Request to see if it is a `HEAD` request.

**Comment by akrabat** [comment]
**Todo:**

1. make sure that the user can detect a HEAD request inside the route callable that's attached to a GET request
2. Ensure that we don't send the request body if it's a HEAD request.

**Comment by akrabat** [comment]
Addressed in #2547

**Comment by akrabat** [comment]
Re-opening - 3.x needs checking.

**Comment by lordrhodos** [comment]
I did some research for the 3.x code. 

> **Todo:**
>
>    1. make sure that the user can detect a HEAD request inside the route callable that's attached to a GET request

As the PSR-7 request is passed to the callable it is no problem to check manually if this is actually a `HEAD` request, e.g.

```php
$app->get('/', function (Request $request, Response $response, array $args) {
...
    $isHeadMethod = $request->getMethod() === 'HEAD';
...
});
```

> 2. Ensure that we don't send the request body if it's a HEAD request.

This is kind of tricky. The mentioned `Response::isEmpty` method just checks for a response status code that should not contain a body:

https://github.com/slimphp/Slim/blob/715d1306dd221f016b2f3f1d132a0ada3c3ac9a8/Slim/Http/Response.php#L376-L382

I think we need to integrate this at `Slim\App::isEmptyResponse` and add a check if we are dealing with a `HEAD` request, which will leave us with another fix that we need to attend here:

https://github.com/slimphp/Slim/blob/715d1306dd221f016b2f3f1d132a0ada3c3ac9a8/Slim/App.php#L617-L619

Here we need to check if this is not a `HEAD` request before returning a response with removed headers.

**Comment by l0gicgate** [comment]
Fixed with #2547 for Slim 4 and #2576 for Slim 3
